content/posts/2025-01-03-packets-xlation-enrichment.md

```diff
@@ -0,0 +1,197 @@
+---
+layout: :theme/post
+title: "Enhancing NetObserv for Kubernetes Service Flows using eBPF"
+description: NetObserv Packet Translation provide enriched service endpoint POD information
+tags: eBPF,Services,Monitoring,Troubleshooting
+authors: [msherif1234]
+---
+
+# Enhancing NetObserv for Kubernetes Service Flows using eBPF: Adding Translated Endpoint Information
+
+## Introduction
+
+In a cloud-native environment, Kubernetes has become the de facto standard for managing containerized workloads.
+While its service discovery and load-balancing features are robust, gaining visibility into the actual endpoints
+serving traffic often requires complex instrumentation or external tooling.
+This is where eBPF (extended Berkeley Packet Filter) shines, offering a way to deeply observe and enrich Kubernetes
+service flows without intrusive changes to your applications.
+
+In this blog, we'll explore how eBPF can be used to add translated endpoint (xlated endpoint) information to
+Kubernetes service flows, providing insights into backend behavior and improving observability.
+
+## The Challenge
+
+When traffic flows through a Kubernetes Service, it often gets abstracted away by kube-proxy or other load balancers.
+For example, a request to `my-service` on `ClusterIP` is transparently routed to one of the available pods.
+However, network observability will capture both the service traffic and the virtual service endpoint as two separate flows.
```

Comment:
Behind the scene it's more than network observability issue here right ?
Any observability tool based on ebpf will report this. Maybe expose that instead

Reply:
not all observability tools are doing flow based capturing like us ? I guess what I meant to say here there is no association between svc traffic and its endpoint traffic user have to do this correlation manually but with new feature correlation is done for them

Reply:
Kubeshark is adding hubs to monitor K8S API for example: https://docs.kubeshark.co/en/anatomy_of_kubeshark
Istio generates telemetry for meshes: https://istio.io/latest/docs/concepts/observability/
Managed kubernetes services like AKS / EKS does their own soup also. Not sure how it's implemented.

The point of view differs so they should be able to correlate / display differently.